pika/channel.py

```diff
@@ -1267,9 +1269,40 @@ def _on_synchronous_complete(self, _method_frame_unused):
         """
         LOGGER.debug('%i blocked frames', len(self._blocked))
         self._blocking = None
+        # self._blocking must be checked here as a callback could
+        # potentially change the state of that variable during an
+        # iteration of the while loop
         while self._blocked and self._blocking is None:
             self._rpc(*self._blocked.popleft())
 
+    def _drain_blocked_methods_on_remote_close(self):
+        """This is called when the broker sends a Channel.Close. It checks the
+        blocked method queue for pending client-initiated Channel.Close methods
+        and ensures their callbacks are processed, but does *not* execute the
+        blocked method
+
+        """
+        LOGGER.debug('Draining %i blocked frames due to remote Channel.Close', len(self._blocked))
+        self._blocking = None
+        # self._blocking must be checked here as a callback could
+        # potentially change the state of that variable during an
+        # iteration of the while loop. Technically, we can only enter
+        # this method once due to the self.is_closing check, but we
+        # will iterate over all blocked methods for debug logging purposes
+        while self._blocked and self._blocking is None:
+            method = self._blocked.popleft()[0]
+            if isinstance(method, spec.Channel.Close):
+                # in this case, the Channel.Close method has not yet been
+                # sent to the broker but the channel has been
+                # closed by the broker by a remote Channel.Close. We do not
+                # want to send this method, but process it as though it had
+                # been sent
```

Comment:
> We do not want to send this method, but process it as though it had been sent

Just FYI: technically, there would be no harm in sending the blocked `Channel.Close` to the server, since AMQP 0.9.1 specification obliges the server to respond with `Close-OK` in this case, while dropping all other incoming methods, save for `Close-OK`

Reply:
That's interesting, because that's not what happens with RabbitMQ. What I saw is `504` error response where RabbitMQ says it expects a `channel.open`. Maybe because in that case the test used the same channel number?